server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java

```diff
@@ -379,7 +379,11 @@ private boolean complete(@Nullable V v, @Nullable Exception e, int finalState) {
             } else if (getState() == COMPLETING) {
                 // If some other thread is currently completing the future, block until
                 // they are done so we can guarantee completion.
-                acquireShared(-1);
+                // Don't use acquire here, to prevent false-positive deadlock detection
+                // when multiple threads from the same pool are completing the future
+                while (isDone() == false) {
+                    Thread.onSpinWait();
+                }
```

Comment:
I imagine we stay in this state extremely briefly as we're just waiting on two assignments, surely busy wait is OK here?

Reply:
We could also use some other synchronisation mechanism here to wait/notify (that didn't add the waiter to the waiters for the `Sync`), but that might be overkill?

Reply:
Seems ok to me; possibly this is even more efficient, not that efficiency particularly matters in these cases.